source/MaterialXContrib/MaterialXNode/MaterialXNode.cpp

```diff
@@ -121,35 +121,55 @@ bool MaterialXNode::getInternalValue(const MPlug& plug, MDataHandle& dataHandle)
 
 bool MaterialXNode::setInternalValue(const MPlug& plug, const MDataHandle& dataHandle)
 {
-    auto displayError = [this](const char* error)
+    auto createAndRegisterFragment = [this]()
     {
-        MString message("Failed to create shader for ");
-        message += typeName();
-        message += " '";
-        message += name();
-        message += "': ";
-        message += error;
-        MGlobal::displayError(message);
-    };
+        try
+        {
+            if (_documentFilePath.length() == 0)
+            {
+                return;
+            }
 
-    auto loadDocument = [this]()
-    {
-        return MaterialXMaya::loadDocument(
-            _documentFilePath.asChar(), Plugin::instance().getLibrarySearchPath()
-        );
-    };
+            if (!_document)
```

Comment:
This is a new member of MaterialXNode. It may be non-null even if the MaterialXData is null, e.g. when the document is valid but the element path isn't. This allows us to avoid the overhead of reloading the document as the element path is being edited. Thus, MaterialXNode represents a VP2 shader fragment successfully generated from a MaterialX element - should we actually rename it to something like **MaterialXFragment**?

Reply:
I don't know if anyone will know what a "fragment" is. Maybe MaterialXShader?

Reply:
Yep, I think MaterialXShader sounds good.

Reply:
But that case "elementIsAShader" should probably be renamed - e.g. to "isSurface". Otherwise it's a class named a shader whose instances may not be shaders :-)

Reply:
Sounds good.